demos/sslecho/main.c

```diff
@@ -0,0 +1,301 @@
+	/* If client get remote server address (could be 127.0.0.1) */
+	if (!isServer) {
+		if (argc != 3) {
+			usage();
+			/* NOTREACHED */
+		}
+		rem_server_ip = argv[2];
+	}
+
+	/* Create context used by both client and server */
+	ssl_ctx = create_context(isServer);
+	/* Configure context with appropriate key files */
+	configure_context(ssl_ctx);
+
+	/* If server */
+	if (isServer) {
+		printf("We are the server on port: %d\n\n", server_port);
+
+		/* Create server socket; will bind with server port and listen */
+		server_skt = create_socket(true);
+
+		/*
+		 * Loop to accept clients.
+		 * Need to implement timeouts on TCP & SSL connect/read functions
+		 * before we can catch a CTRL-C and kill the server.
+		 */
+		while (server_running) {
+			/* Wait for TCP connection from client */
+			client_skt = accept(server_skt, (struct sockaddr*) &addr,
+					&addr_len);
+			if (client_skt < 0) {
+				perror("Unable to accept");
+				exit(EXIT_FAILURE);
+			}
+
+			printf("Client TCP connection accepted\n");
+
+			/* Create server SSL structure using the new socket */
+			ssl = SSL_new(ssl_ctx);
+			SSL_set_fd(ssl, client_skt);
+
+			/* Wait for SSL connection from the client */
+			if (SSL_accept(ssl) <= 0) {
+				ERR_print_errors_fp(stderr);
+			} else {
+
+				printf("Client SSL connection accepted\n\n");
+
+				/* Echo loop */
+				while (true) {
+					/* Get message from client; will fail if client closes connection */
+					if ((rxlen = SSL_read(ssl, rxbuf, rxcap)) <= 0) {
+						if (rxlen == 0) {
+							printf("Client closed connection\n");
+						}
+						ERR_print_errors_fp(stderr);
+						break;
+					}
+					/* Insure null terminated input */
+					rxbuf[rxlen] = 0;
+					/* Look for kill switch */
+					if (strcmp(rxbuf, "kill\n") == 0) {
+						/* Terminate...with extreme prejudice */
+						printf("Server received 'kill' command\n");
+						server_running = false;
+						break;
+					}
+					/* Show received message */
+					printf("Received: %s", rxbuf);
+					/* Echo it back */
+					if (SSL_write(ssl, rxbuf, rxlen) <= 0) {
+						ERR_print_errors_fp(stderr);
+					}
+				}
+			}
+			if (server_running) {
+				/* Cleanup for next client */
+				SSL_shutdown(ssl);
+				SSL_free(ssl);
+				close(client_skt);
+			}
+		}
+		printf("Server exiting...\n");
+	}
+	/* Else client */
+	else {
+		printf("We are the client\n\n");
+
+		/* Create "bare" socket */
+		client_skt = create_socket(false);
+		/* Set up connect address */
+		addr.sin_family = AF_INET;
+		inet_pton(AF_INET, rem_server_ip, &addr.sin_addr.s_addr);
+		addr.sin_port = htons(server_port);
+		/* Do TCP connect with server */
+		if (connect(client_skt, (struct sockaddr*) &addr, sizeof(addr)) != 0) {
+			perror("Unable to TCP connect to server");
+			goto Exit;
+		} else {
+			printf("TCP connection to server successful\n");
+		}
+
+		/* Create server SSL structure using client socket */
+		ssl = SSL_new(ssl_ctx);
+		SSL_set_fd(ssl, client_skt);
+
```

Comment:
In addition to `SSL_set_tlsext_host_name()` we should also call `SSL_set1_host` here. This instructs OpenSSL to verify that the hostname in the certificate matches what we expect and (in combination with the `SSL_CTX_set_verify` function above) will cause OpenSSL to abort if it doesn't.

This does also mean that the certificate must include the hostname in the CN of the Subject (or in a SAN field) in the certificate (which it doesn't at the moment in your example cert).

Reply:
Oddly, SSL_set_tlsext_host_name() and SSL_set1_host() were added without changing the cert file, and it seems to work...
Will commit shortly.